subsystem/src/main/java/org/wildfly/subsystem/resource/capability/CapabilityReferenceAttributeDefinition.java

```diff
@@ -0,0 +1,86 @@
+import org.jboss.as.controller.AttributeParser;
+import org.jboss.as.controller.OperationContext;
+import org.jboss.as.controller.OperationFailedException;
+import org.jboss.as.controller.SimpleAttributeDefinition;
+import org.jboss.as.controller.operations.validation.StringLengthValidator;
+import org.jboss.as.controller.registry.AttributeAccess.Flag;
+import org.jboss.dmr.ModelNode;
+import org.jboss.dmr.ModelType;
+import org.wildfly.subsystem.resource.SimpleResource;
+import org.wildfly.subsystem.resource.ResourceModelResolver;
+import org.wildfly.subsystem.service.ServiceDependency;
+
+/**
+ * An attribute definition referencing some capability.
+ * Resolves directly to a {@link ServiceDependency} via {@link #resolve(OperationContext, ModelNode)}.
+ */
+public class CapabilityReferenceAttributeDefinition<T> extends SimpleAttributeDefinition implements ResourceModelResolver<ServiceDependency<T>> {
+
+    private final CapabilityReferenceResolver<T> resolver;
+
+    CapabilityReferenceAttributeDefinition(Builder<T> builder) {
+        super(builder);
+        this.resolver = builder.resolver;
+    }
+
+    @Override
+    public ServiceDependency<T> resolve(OperationContext context, ModelNode model) throws OperationFailedException {
+        String value = this.resolveModelAttribute(context, model).asStringOrNull();
+        Map.Entry<String, String[]> resolved = this.resolver.resolve(context, new SimpleResource(model), value);
+        return (resolved != null) ? ServiceDependency.on(resolved.getKey(), this.resolver.getRequirement().getType(), resolved.getValue()) : ServiceDependency.of(null);
+    }
+
+    public static class Builder<T> extends AbstractAttributeDefinitionBuilder<Builder<T>, CapabilityReferenceAttributeDefinition<T>> {
+
+        final CapabilityReferenceResolver<T> resolver;
+
+        public Builder(String attributeName, CapabilityReference<T> reference) {
+            super(attributeName, ModelType.STRING);
+            // Capability references never allow expressions
+            this.setAllowExpression(false);
+            this.setAttributeParser(AttributeParser.SIMPLE);
+            this.setCapabilityReference(reference);
+            this.setFlags(Flag.RESTART_RESOURCE_SERVICES);
+            this.setValidator(new StringLengthValidator(1));
+            this.resolver = reference;
+        }
+
+        public Builder(String attributeName, CapabilityReferenceAttributeDefinition<T> basis) {
+            super(attributeName, basis);
+            this.resolver = basis.resolver;
+        }
+
+        @Override
+        public Builder<T> setDefaultValue(ModelNode defaultValue) {
+            // A capability reference must not specify a default value
+            if ((defaultValue != null) && defaultValue.isDefined()) {
+                throw new UnsupportedOperationException();
+            }
+            return this;
+        }
+
+        @Override
+        public Builder<T> setAllowExpression(boolean allowExpression) {
```

Comment:
@pferraro This should be explained in javadoc.

Reply:
This meaning L76 :)

Reply:
Fixed.